spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java

```diff
@@ -466,6 +471,163 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+            .rewriteIf(manifest -> true)
+            .sort(manifestSortKeys)
+            .option(RewriteManifestsSparkAction.USE_CACHING, useCaching)
+            .execute();
+
+    table.refresh();
+    List<ManifestFile> newManifests = table.currentSnapshot().allManifests(table.io());
+
+    assertThat(result.rewrittenManifests()).hasSize(1);
+    assertThat(result.addedManifests()).hasSizeGreaterThanOrEqualTo(2);
+
+    assertThat(newManifests).hasSizeGreaterThanOrEqualTo(2);
+
+    // Rewritten manifests are clustered by c3_bucket - each should contain only a subset of the
+    // lower and upper bounds
+    // of the partition 'c3'.
+    List<Pair<Integer, Integer>> c3Boundaries =
+        newManifests.stream()
+            .map(manifest -> manifest.partitions().get(2))
+            .sorted(
+                Comparator.comparing(
+                    ptn -> Conversions.fromByteBuffer(Types.IntegerType.get(), ptn.lowerBound())))
+            .map(
+                p ->
+                    Pair.of(
+                        (Integer)
+                            Conversions.fromByteBuffer(Types.IntegerType.get(), p.lowerBound()),
+                        (Integer)
+                            Conversions.fromByteBuffer(Types.IntegerType.get(), p.upperBound())))
+            .collect(Collectors.toList());
+
+    List<Integer> lowers = c3Boundaries.stream().map(t -> t.first()).collect(Collectors.toList());
+    List<Integer> uppers = c3Boundaries.stream().map(t -> t.second()).collect(Collectors.toList());
+
+    // With custom sorting, this looks like
+    // - manifest 1 -> [lower bound = 0, upper bound = 4]
+    // - manifest 2 -> [lower bound = 4, upper bound = 9]
+    // Without the custom sorting, each manifest tracks the full range of c3 upper/lower bounds.
+    // AKA they look like
+    // - manifest 1 -> [lower bound = 0, upper bound = 9]
+    // - manifest 2 -> [lower bound = 0, upper bound = 9]
+    // So the upper bound of the partitions tracked by the first file should be LEQ the lower bounds
+    // of the second. Etc
+    assertThat(uppers.get(0))
+        .withFailMessage(
+            "Upper bound of first manifest partition should be LEQ lower bound of second")
+        .isLessThanOrEqualTo(lowers.get(1));
+
+    // Each file should contain less than the full c3 partition span
+    c3Boundaries.forEach(
+        boundary -> {
+          assertThat(boundary.second() - boundary.first())
+              .withFailMessage(
+                  "Manifest should contain less than the full range of c3 bucket partitions")
+              .isLessThanOrEqualTo(c3PartitionMax - c3PartitionMin);
+        });
+
+    // c3's Bucket(10) partition means our true lower bound = 0 and true upper bound is 9. The first
+    // manifest should
+    // include the lower bound of 0, and the last should have the upper bound of 9
+    assertThat(lowers.get(0))
+        .withFailMessage("Lower bound of first manifest partition should be 0")
```

Comment:
same as above

Reply:
unresolving this comment since it hasn't been updated yet

Reply:
My bad, missed these. Fixed all occurrences.